File: src/api/neural-queue.ts
/**
 * Tauri API wrapper for the SuperMemo neural queue (supermemo-faithful-queue
 * Phase 4) — the optional "Go neural" creative-exploration mode.
 *
 * The neural queue is layered on top of the priority queue: entering neural
 * review builds a queue by spreading activation from a seed element; exiting
 * returns to the priority queue without mutating it. Normal learning never
 * touches these calls.
 */

import { invokeCommand } from "../lib/tauri";
import type { EmbeddingConfig } from "./ai-learning";

export type NeuralElementKind = "document" | "extract" | "learning_item";

/** A persisted row in the neural_queue table. */
export interface NeuralQueueRow {
  element_id: number;
  position: number;
  priority_value: number;
  consumed: boolean;
  updated_at: string;
}

/**
 * Enter neural review (SuperMemo's *Learn : Go neural*): build the neural
 * queue by spreading activation seeded at the given element. Returns the
 * number of elements queued. The priority queue is not mutated.
 *
 * When `embeddingConfig` is supplied, the spreading activation also follows
 * embedding-similarity edges (semantic neighbors via the RAG chunk embeddings).
 * Omit it for collections that aren't indexed — only the tree-topology
 * relationships then drive the queue.
 */
export async function buildNeuralQueue(
  elementKind: NeuralElementKind,
  elementRefId: string,
  embeddingConfig?: EmbeddingConfig,
): Promise<number> {
  return invokeCommand<number>("build_neural_queue", {
    elementKind,
    elementRefId,
    embeddingConfig: embeddingConfig ?? null,
  });
}

/** Fetch up to `limit` unconsumed elements from the front of the neural queue. */
export async function getNeuralQueueFront(limit: number = 20): Promise<NeuralQueueRow[]> {
  return invokeCommand<NeuralQueueRow[]>("get_neural_queue_front", { limit });
}

/**
 * A neural-queue front entry JOINed with its element_tree identity, so the UI
 * can resolve each position back to the concrete document / extract / card to
 * render. `element_id` alone is unrenderable without this JOIN.
 */
export interface ResolvedNeuralQueueEntry {
  element_id: number;
  position: number;
  priority_value: number;
  /** `"document"` | `"extract"` | `"learning_item"`. */
  element_kind: NeuralElementKind;
  /** The concrete item's uuid (documents.id / extracts.id / learning_items.id). */
  element_ref_id: string;
}

/**
 * Fetch up to `limit` unconsumed neural-queue entries, each resolved to its
 * concrete element identity. This is what the "Go neural" review UI calls —
 * unlike `getNeuralQueueFront`, the returned entries carry `element_kind` +
 * `element_ref_id` so the frontend can render the actual document / extract /
 * card. Entries arrive in presentation order (ascending `position`).
 */
export async function getNeuralQueueResolvedFront(
  limit: number = 60,
): Promise<ResolvedNeuralQueueEntry[]> {
  return invokeCommand<ResolvedNeuralQueueEntry[]>("get_neural_queue_resolved_front", { limit });
}

/** Mark `elementId` as studied (consumed). Returns true if a row was updated. */
export async function consumeNeuralQueueElement(elementId: number): Promise<boolean> {
  return invokeCommand<boolean>("consume_neural_queue_element", { elementId });
}

/**
 * Refill the neural queue by spreading activation seeded at the given element,
 * but **only if** the remaining count is below the depletion threshold (< 20).
 * Returns the new queue size if a refill ran, or null if the threshold was met.
 * This is the depletion trigger — call it after `consumeNeuralQueueElement`
 * when the user finishes studying an element in neural review.
 *
 * Pass the same `embeddingConfig` used at build time so semantic neighbors
 * remain in play on refill.
 */
export async function refillNeuralQueueIfDepleted(
  elementKind: NeuralElementKind,
  elementRefId: string,
  embeddingConfig?: EmbeddingConfig,
): Promise<number | null> {
  return invokeCommand<number | null>("refill_neural_queue_if_depleted", {
    elementKind,
    elementRefId,
    embeddingConfig: embeddingConfig ?? null,
  });
}

/** The number of unconsumed elements remaining in the neural queue. */
export async function getNeuralQueueRemaining(): Promise<number> {
  return invokeCommand<number>("get_neural_queue_remaining");
}

File: src-tauri/src/commands/neural_queue.rs
//! Tauri commands for the SuperMemo neural queue (supermemo-faithful-queue
//! Phase 4) — the optional "Go neural" creative-exploration mode.
//!
//! These commands are the entry points the frontend invokes to enter neural
//! review, fetch the next elements, mark them studied, and let the depletion
//! trigger refill the queue. Normal learning never touches these; the neural
//! queue is layered on top of the priority queue and does not mutate it.

use tauri::State;

use crate::database::{ElementKind, ElementTreeRepository, NeuralQueueRepository, Repository};
use crate::error::{IncrementumError, Result};

/// Enter neural review (SuperMemo's *Learn : Go neural*): build the neural
/// queue by spreading activation seeded at the element derived from
/// `(element_kind, element_ref_id)`. Returns the number of elements queued.
///
/// The priority queue is **not** mutated — only read for intrinsic priorities.
#[tauri::command]
pub async fn build_neural_queue(
    element_kind: String,
    element_ref_id: String,
    repo: State<'_, Repository>,
) -> Result<usize> {
    let kind = match element_kind.as_str() {
        "document" => ElementKind::Document,
        "extract" => ElementKind::Extract,
        "learning_item" => ElementKind::LearningItem,
        other => {
            return Err(IncrementumError::InvalidInput(format!(
                "unknown element_kind '{other}'"
            )))
        }
    };
    let et = ElementTreeRepository::new(repo.pool().clone());
    let seed = et
        .find_node_id(kind, &element_ref_id)
        .await?
        .ok_or_else(|| {
            IncrementumError::NotFound(format!(
                "element_tree node for {element_kind}:{element_ref_id}"
            ))
        })?;
    let nqr = NeuralQueueRepository::new(repo.pool().clone());
    let entries = nqr.build(seed).await?;
    Ok(entries.len())
}

/// Fetch up to `limit` unconsumed elements from the front of the neural queue,
/// in presentation order.
#[tauri::command]
pub async fn get_neural_queue_front(
    limit: Option<usize>,
    repo: State<'_, Repository>,
) -> Result<Vec<crate::database::NeuralQueueRow>> {
    let nqr = NeuralQueueRepository::new(repo.pool().clone());
    nqr.front(limit.unwrap_or(20)).await
}

/// Mark `element_id` as studied (consumed). If the queue depletes below the
/// threshold as a result, a refill is **not** triggered here — the frontend
/// calls `refill_neural_queue_if_depleted` after consuming, seeded at the
/// element just studied (task 4.11 depletion trigger).
#[tauri::command]
pub async fn consume_neural_queue_element(
    element_id: i64,
    repo: State<'_, Repository>,
) -> Result<bool> {
    let nqr = NeuralQueueRepository::new(repo.pool().clone());
    nqr.consume(element_id).await
}

/// Refill the neural queue by spreading activation seeded at the given element
/// **only if** the remaining count is below the depletion threshold. Returns
/// Some(count) if a refill ran, None if the threshold was met. This is the
/// depletion trigger (task 4.11): it runs after a consume when remaining < 20,
/// not on every grade.
#[tauri::command]
pub async fn refill_neural_queue_if_depleted(
    element_kind: String,
    element_ref_id: String,
    repo: State<'_, Repository>,
) -> Result<Option<usize>> {
    let kind = match element_kind.as_str() {
        "document" => ElementKind::Document,
        "extract" => ElementKind::Extract,
        "learning_item" => ElementKind::LearningItem,
        other => {
            return Err(IncrementumError::InvalidInput(format!(
                "unknown element_kind '{other}'"
            )))
        }
    };
    let nqr = NeuralQueueRepository::new(repo.pool().clone());
    if !nqr.needs_refill().await? {
        return Ok(None);
    }
    let et = ElementTreeRepository::new(repo.pool().clone());
    let seed = et
        .find_node_id(kind, &element_ref_id)
        .await?
        .ok_or_else(|| {
            IncrementumError::NotFound(format!(
                "element_tree node for {element_kind}:{element_ref_id}"
            ))
        })?;
    let entries = nqr.build(seed).await?;
    Ok(Some(entries.len()))
}

/// The number of unconsumed elements remaining in the neural queue.
#[tauri::command]
pub async fn get_neural_queue_remaining(repo: State<'_, Repository>) -> Result<usize> {
    let nqr = NeuralQueueRepository::new(repo.pool().clone());
    nqr.remaining().await
}
